### PR TITLE
Parallel testing support

### PR DESCRIPTION
**Pull request by lukeraymonddowning**
| Q             | A
| ------------- | ---
| Bug fix?      | no
| New feature?  | yes

Hi all,

Here it is! A very early draft of parallel testing in Pest. This uses brianium/paratest as the base runner.

I'll be adding more detail here as I continue work on this, but wanted to get it out in the open for suggestions and changes.

Kind Regards,
Luke

### CODE REVIEW

**Comment by lukeraymonddowning** [comment]
Currently, we're using the `SuiteLoader` to fill out Pest's test case, then using that to merge and run the files. It might be better for us to load the files in directly instead.

**Comment by lukeraymonddowning** [comment]
So, there is one breaking test in the repo. That test isn't related to parallel and the bug can be reproduced without it:

```php
bin/pest tests/PHPUnit/CustomTestCaseInSubFolders/SubFolder/SubFolder/UsesPerSubDirectory.php
```

will reproduce outside of this branch.

**Comment by lukeraymonddowning** [comment]
Todos (in no particular order):

- [ ] PR some changes to Paratest itself which would negate having to override entire classes for 1 or 2 lines
- [ ] Build a custom printer that supports the additional Paratest functionality but outputs Pest style
- [ ] Allow the `--coverage` flag, which would run paratest's coverage functionality then output Pest style
- [ ] Test edge cases and other operating systems (Windows, Linux)

**Comment by owenvoke** [comment]
> So, there is one breaking test in the repo. That test isn't related to parallel and the bug can be reproduced without it:
> 
> ```
> bin/pest tests/PHPUnit/CustomTestCaseInSubFolders/SubFolder/SubFolder/UsesPerSubDirectory.php
> ```
> 
> will reproduce outside of this branch.

I wonder if that's because the required `uses()` calls in `tests/Pest.php` aren't being called or something for a single test run? 🤔

**Comment by lukeraymonddowning** [comment]
@owenvoke its actually the Pest file in tests/PHPUnit/Pest.php that isn't being called.

**Comment by owenvoke** [comment]
@lukeraymonddowning, sorry yes. I corrected my comment earlier. 👍🏻 Forgot where the nested file was.

**Comment by nuernbergerA** [comment]
* [x] Allow the --coverage flag, which would run paratest's coverage functionality then output Pest style

File: src/Actions/InteractsWithPlugins.php [hunk 1]
```diff
@@ -0,0 +1,50 @@
...
+
+/**
+ * @internal
+ */
+final class InteractsWithPlugins
+{
+    /**
+     * Transform the input arguments by passing it to the relevant plugins.
+     *
+     * @param array<string> $argv
```
[the remark this answers is not in the document]
Comment (lukeraymonddowning) [reply src/Actions/InteractsWithPlugins.php]: Added

File: src/Actions/InteractsWithPlugins.php [hunk 2]
```diff
@@ -0,0 +1,50 @@
...
+     *
+     * @return array<string>
```
[the remark this answers is not in the document]
Comment (lukeraymonddowning) [reply src/Actions/InteractsWithPlugins.php]: Added

File: bin/pest [hunk 3]
```diff
@@ -51,5 +53,11 @@ use Symfony\Component\Console\Output\OutputInterface;
         }
     }
 
+    $_SERVER['argv'] = InteractsWithPlugins::handleArguments($_SERVER['argv']);
+
+    if ($argv->hasParameterOption('--parallel')) {
+        exit(ParaTestCommand::applicationFactory($testSuite->rootPath)->run(new ArgvInput()));
```
[the remark this answers is not in the document]
Comment (lukeraymonddowning) [reply bin/pest]: Refactored:

1) Plugin now has its own `Command` class, which is called instead of Paratest
2) We no longer morph the `$_SERVER['argv']` input at all, but rather transform an array

File: bin/pest [hunk 4]
```diff
@@ -2,6 +2,8 @@
 <?php declare(strict_types=1);
 
 use NunoMaduro\Collision\Provider;
+use ParaTest\Console\Commands\ParaTestCommand;
```
[the remark this answers is not in the document]
Comment (lukeraymonddowning) [reply bin/pest]: Removed

File: src/Concerns/Testable.php [hunk 5]
```diff
@@ -304,4 +304,12 @@ public function getPrintableTestCaseName(): string
     {
         return ltrim(self::class, 'P\\');
     }
+
+    /**
+     * Determine whether this test case is being executed in a parallel environment.
+     */
+    public function isInParallel(): bool
```
[the remark this answers is not in the document]
Comment (lukeraymonddowning) [reply src/Concerns/Testable.php]: Removed

File: tests/Hooks/BeforeAllTest.php [hunk 6]
```diff
@@ -1,13 +1,14 @@
...
 use Pest\Support\Str;
+use Pest\TestSuite;
 
 global $globalHook;
 
 // HACK: we have to determine our $globalHook->calls baseline. This is because
 // two other tests are executed before this one due to filename ordering.
 $args   = $_SERVER['argv'] ?? [];
-$single = isset($args[1]) && Str::endsWith(__FILE__, $args[1]);
+$single = (isset($args[1]) && Str::endsWith(__FILE__, $args[1])) || TestSuite::getInstance()->isInParallel;
```
[the remark this answers is not in the document]
Comment (lukeraymonddowning) [reply tests/Hooks/BeforeAllTest.php]: Required as discussed yesterday. Altered to use the new `$_SERVER['PEST_PARALLEL']`

File: tests/PHPUnit/CustomTestCaseInSubFolders/SubFolder/SubFolder/CustomTestCaseInSubFolder.php [hunk 7]
```diff
@@ -2,7 +2,7 @@
 
 declare(strict_types=1);
 
-namespace Tests\SubFolder\SubFolder\SubFolder;
+namespace Tests\CustomTestCaseInSubFolders\SubFolder\SubFolder;
```
[the remark this answers is not in the document]
Comment (lukeraymonddowning) [reply tests/PHPUnit/CustomTestCaseInSubFolders/SubFolder/SubFolder/CustomTestCaseInSubFolder.php:5]: Just thinking, isn't this actually correct? Before, the namespace didn't match the folder structure...

File: tests/PHPUnit/Pest.php [hunk 8]
```diff
@@ -1,3 +1,5 @@
 <?php
 
-uses(Tests\SubFolder\SubFolder\SubFolder\CustomTestCaseInSubFolder::class)->in('CustomTestCaseInSubFolders/SubFolder');
+use Tests\CustomTestCaseInSubFolders\SubFolder\SubFolder\CustomTestCaseInSubFolder;
```
[the remark this answers is not in the document]
Comment (lukeraymonddowning) [reply tests/PHPUnit/Pest.php]: Moved

File: tests/PHPUnit/CustomTestCaseInSubFolders/SubFolder/SubFolder/UsesPerSubDirectory.php [hunk 9]
```diff
@@ -1,5 +1,7 @@
 <?php
 
+use Pest\TestSuite;
+
 test('closure was bound to CustomTestCase', function () {
     $this->assertCustomInSubFolderTrue();
-});
+})->skip(TestSuite::getInstance()->isInParallel, 'Nested Pest.php files are not loaded in parallel.');
```
[the remark this answers is not in the document]
Comment (lukeraymonddowning) [reply tests/PHPUnit/CustomTestCaseInSubFolders/SubFolder/SubFolder/UsesPerSubDirectory.php]: Moved

File: composer.json [hunk 10]
```diff
@@ -43,7 +43,8 @@
         "illuminate/console": "^8.47.0",
         "illuminate/support": "^8.47.0",
         "laravel/dusk": "^6.15.0",
-        "pestphp/pest-dev-tools": "dev-master"
+        "pestphp/pest-dev-tools": "dev-master",
+        "pestphp/pest-plugin-parallel": "dev-master"
```
[the remark this answers is not in the document]
Comment (lukeraymonddowning) [reply composer.json]: Done

File: src/Functions.php [hunk 11]
```diff
@@ -2,6 +2,7 @@
 
 declare(strict_types=1);
 
+use Pest\Concerns\Testable;
```
[the remark this answers is not in the document]
Comment (lukeraymonddowning) [reply src/Functions.php]: Removed

File: src/Functions.php [hunk 12]
```diff
@@ -85,7 +86,7 @@ function uses(string ...$classAndTraits): UsesCall
      * is the test description; the second argument is
      * a closure that contains the test expectations.
      *
-     * @return TestCall|TestCase|mixed
+     * @return TestCall|TestCase|Testable|mixed
```
[the remark this answers is not in the document]
Comment (lukeraymonddowning) [reply src/Functions.php]: Removed

File: src/Repositories/TestRepository.php [hunk 13]
```diff
@@ -29,7 +29,7 @@ final class TestRepository
     /**
      * @var array<string, TestCaseFactory>
      */
-    private $state = [];
+    public $state = [];
```
[the remark this answers is not in the document]
Comment (lukeraymonddowning) [reply src/Repositories/TestRepository.php]: Updated

File: src/TestSuite.php [hunk 14]
```diff
@@ -73,6 +73,13 @@ final class TestSuite
      */
     public $testPath;
 
+    /**
+     * Whether this test is running as part of a parallel suite.
+     *
+     * @var bool
+     */
+    public $isInParallel = false;
```
[the remark this answers is not in the document]
Comment (lukeraymonddowning) [reply src/TestSuite.php]: Removed

File: tests/Visual/Success.php [hunk 15]
```diff
@@ -35,4 +37,5 @@
         expect(implode("\n", $output))->toContain(file_get_contents($snapshot));
     }
 })->skip(!getenv('REBUILD_SNAPSHOTS') && getenv('EXCLUDE'))
+    ->skip(TestSuite::getInstance()->isInParallel)
```
[the remark this answers is not in the document]
Comment (lukeraymonddowning) [reply tests/Visual/Success.php]: All working now 🎉

File: composer.json [hunk 16]
```diff
@@ -84,5 +85,11 @@
                 "Pest\\Laravel\\PestServiceProvider"
             ]
         }
-    }
+    },
+    "repositories": [
```
[the remark this answers is not in the document]
Comment (lukeraymonddowning) [reply composer.json]: Removed

**Review by nunomaduro — APPROVED** [review]

**Review by nunomaduro — APPROVED** [review]
